cylc/flow/etc/job.sh

```diff
@@ -121,11 +119,23 @@ cylc__job__main() {
     mkdir -p "$(dirname "${CYLC_TASK_WORK_DIR}")" || true
     mkdir -p "${CYLC_TASK_WORK_DIR}"
     cd "${CYLC_TASK_WORK_DIR}"
-    # User Environment, Pre-Script, Script and Post-Script
-    typeset func_name=
-    for func_name in 'user_env' 'pre_script' 'script' 'post_script'; do
-        cylc__job__run_inst_func "${func_name}"
-    done
+    # Env-Script, User Environment, Pre-Script, Script and Post-Script
+    cylc__job__run_user_scripts &
+    CYLC_TASK_USER_SCRIPT_PID=$!
+    wait "${CYLC_TASK_USER_SCRIPT_PID}" || {
+        # Check return code for signals (value greater than 128).
+        typeset ret_code="$?"
+        if ((ret_code > 128)); then
+            # Bypass ERR trap and get on with the signal (if we received one)
+            # or EXIT trap (if the subshell or one of its children received
+            # a signal, e.g. SIGPIPE in a command pipeline).
+            exit "$ret_code"
+        else
+            # Trigger ERR trap while preserving the exit code (return won't do).
+            (exit "$ret_code")
+            return
+        fi
+    }
```

Comment:
This is a pretty crafty solution but it's necessary to use a backgrounded function rather than subshell to preserve the error code on bash 4.2.